### PR TITLE
Add ability to store states into etcd directly

### PR DESCRIPTION
**Pull request by xenolog**
Now Netchecker server is able to work with either of the following storage backends:
- etcd
- k8s TPR

TPR becomes deprecated in Kubernetes (CRD is the substitution for TPR) and will not be supported in Netchecker in future.
etcd becomes the default persistent storage for Netchecker.
This implementation uses etcd feature of setting TTL for data that is being stored. So that outdated data will be deleted automatically and Netchecker application will always have only relevant agents data in the storage. This way, determining of agent report obsolescence is simply a check of presence of the data for particular agent in the storage. etcd is now responsible for ensuring correct timing (i.e., removing outdated data according to TTL) and RAFT (which is in use in etcd) helps here if case of nodes dates difference.

Closes: https://github.com/Mirantis/k8s-netchecker-server/issues/80

### CODE REVIEW

File: pkg/utils/storer_etcd.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+		glog.Errorf("Can't fetch tree '%s' recursively from etcd: %v", dirName, err)
+		return agentsData
+	}
+
+	// Iterate to nodes
+	for _, node := range resp.Node.Nodes {
+		npath := strings.Split(node.Key, "/")
+		nname := npath[len(npath)-1]
+		max_AgentSpec.Uptime = 0
+		//glog.Info(nname, len(node.Nodes))
```
Comment (AlexeyKasatkin) [inline pkg/utils/storer_etcd.go]: comment

File: pkg/utils/storer_etcd.go [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+	ProcessResponse(rw, agentData)
+}
+
+func (h *EtcdAgentStorage) CheckAgents() ([]string, []string, error) {
+
+	absent := []string{}
+	outdated := []string{}
+
+	// pods, err := h.KubeClient.Pods()
```
Comment (AlexeyKasatkin) [inline pkg/utils/storer_etcd.go]: ?

File: pkg/utils/storer_k8s.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+	agentData.LastUpdated = time.Now()
+	glog.V(10).Infof("Updating the agents resource with value: %v", agentData)
+
+	agentName := rp.ByName("name")
+	// ags := h.ExtensionsClientset.Agents()
+	// glog.Infoln("*** ags:", ags)
+	// agsl, _ := ags.List()
+	// glog.Infoln("*** ags list:", agsl)
+	// _, err = ags.Get(agentName)
```
Comment (AlexeyKasatkin) [inline pkg/utils/storer_k8s.go]: ?

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -91,6 +90,7 @@ func ProcessRequest(r *http.Request, dst interface{}, rw http.ResponseWriter) er
 	ep := &errProcessor{}
 	body := ep.ReadBody(r)
 	ep.UnmarshalBytes(body, dst)
+	// glog.Errorf(">>>%s<<<", body)
```
Comment (AlexeyKasatkin) [inline pkg/utils/utils.go]: comment

File: pkg/utils/config.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	}
+	return rv
+}
+
+func GetOrCreateConfig() *AppConfig {
+	if !main_config.configured {
+		main_config.PingTimeout = 5 * time.Second
+		main_config.ReportTimeout = 10 * time.Second
+		main_config.ReportTTL = 5 * time.Minute
+		main_config.ReportInterval = 20 * time.Second
```
Comment (AlexeyKasatkin) [inline pkg/utils/config.go]: this should not be hardcoded, I suppose

File: Makefile [hunk 6]
```diff
@@ -141,7 +143,7 @@ $(VENDOR_DIR):
 
 $(BUILD_DIR)/server: $(BUILD_DIR) $(VENDOR_DIR)
 	$(DOCKER_EXEC) bash -xc '$(DOCKER_DEPS) \
-		CGO_ENABLED=0 go build --ldflags "-s -w" \
+		CGO_ENABLED=0 go build --ldflags "-s -w -X main.version=${VERSION}" \
```
Comment (xenolog) [inline Makefile]: Just information, stored into binary file. IT is very helpful while deployment and testing

File: cmd/server.go [hunk 7]
```diff
@@ -22,21 +22,26 @@ import (
...
 func main() {
-	var endpoint string
-	var initKubeProxy bool
-	flag.StringVar(&endpoint, "endpoint", "0.0.0.0:8081", "End point (IP address, port) for server to listen on")
-	flag.BoolVar(&initKubeProxy, "kubeproxyinit", false, "Control initialization kubernetes client set for connectivity check")
+	config := utils.GetOrCreateConfig()
+
+	flag.StringVar(&config.HttpListen, "endpoint", "0.0.0.0:8081", "End point (IP address, port) for server to listen on")
+	flag.BoolVar(&config.UseKubeClient, "kubeproxyinit", false, "Control initialization kubernetes client set for connectivity check")
+	flag.StringVar(&config.EtcdEndpoints, "etcd-endpoints", "", "Etcd endpoints list for store states into etcd instead k8s 3d-party resources")
```
Comment (xenolog) [inline cmd/server.go]: Yes, see examples/deploy.sh

File: cmd/server.go [hunk 8]
```diff
@@ -22,21 +22,26 @@ import (
...
+	flag.StringVar(&config.EtcdTree, "etcd-tree", "netchecker", "Root of tree into Etcd")
```
Comment (xenolog) [inline cmd/server.go]: may be. I thing this ability is helpful while developing

File: pkg/utils/config.go [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	sync.Mutex            // extend for ensures atomic writes; protects the following fields
+	configured     bool   // flag whether App configured
+	UseKubeClient  bool   //
+	EtcdEndpoints  string // how to connect to etcd
+	EtcdTree       string // Root of NetChecker tree into etcd
+	HttpListen     string // IPaddress:PORT for HTTP control API responder
+	PingTimeout    time.Duration
+	ReportTimeout  time.Duration
+	ReportTTL      time.Duration
+	ReportInterval time.Duration
```
Comment (xenolog) [inline pkg/utils/config.go]: fixed

File: pkg/utils/config.go [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+	return rv, err
+}
+
+func (c *AppConfig) ToYaml() ([]byte, error) {
+	var (
+		rv  []byte
+		err error
+	)
+	if rv, err = yaml.Marshal(c); err != nil {
+		//rv = []
```
Comment (xenolog) [inline pkg/utils/config.go]: fixed

File: pkg/utils/config.go [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+		glog.Errorln(err.Error())
+	}
+	return rv, err
+}
+
+func (c *AppConfig) String() string {
+	var (
+		rv  string
+		brv []byte
```
Comment (xenolog) [inline pkg/utils/config.go]: it's no matter for such shot method

File: pkg/utils/storer_etcd.go (hunk 1 above)
Comment (xenolog) [inline pkg/utils/storer_etcd.go]: fixed

File: pkg/utils/storer_k8s.go (hunk 3 above)
Comment (xenolog) [inline pkg/utils/storer_k8s.go]: removed

File: pkg/utils/utils.go (hunk 4 above)
Comment (xenolog) [inline pkg/utils/utils.go]: removed

File: pkg/utils/config.go [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+		rv  []byte
+		err error
+	)
+	if rv, err = yaml.Marshal(c); err != nil {
+		glog.Errorln(err.Error())
+	}
+	return rv, err
+}
+
+func (c *AppConfig) String() string {
```
Comment (AlexeyKasatkin) [inline pkg/utils/config.go]: where do you use it?

**Pull request merged**